jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/openshift/RouteEnricher.java

```diff
@@ -157,58 +151,111 @@ private boolean hasExactlyOneServicePort(ServiceBuilder service, String id) {
             String name = serviceMetadata.getName();
-            if (!hasRoute(listBuilder, name)) {
-                if (StringUtils.isNotBlank(routeDomainPostfix)) {
-                    routeDomainPostfix = prepareHostForRoute(routeDomainPostfix, name);
-                } else {
-                    routeDomainPostfix = "";
+            updateRouteDomainPostFixBasedOnServiceName(name);
+            Route opinionatedRoute = createOpinionatedRouteFromService(serviceBuilder, routeDomainPostfix);
+            if (opinionatedRoute != null) {
+                int routeFromFragmentIndex = getRouteIndexWithName(listBuilder, name);
+                if (routeFromFragmentIndex > 0) { // Merge fragment with Opinionated Route
+                    Route routeFragment = (Route) listBuilder.buildItems().get(routeFromFragmentIndex);
+                    Route mergedRoute = mergeRoute(routeFragment, opinionatedRoute);
+                    KubernetesResourceUtil.removeItemFromKubernetesBuilder(listBuilder, listBuilder.getItems().get(routeFromFragmentIndex));
+                    listBuilder.addToItems(mergedRoute);
+                } else { // No fragment provided. Use Opinionated Route.
+                    listBuilder.addToItems(opinionatedRoute);
                 }
+            }
+        }
+    }
 
-                RoutePort routePort = createRoutePort(serviceBuilder);
-                if (routePort != null) {
-                    RouteBuilder routeBuilder = new RouteBuilder().
-                            withMetadata(serviceMetadata).
-                            withNewSpec().
-                            withPort(routePort).
-                            withNewTo().withKind("Service").withName(name).endTo().
-                            withHost(routeDomainPostfix.isEmpty() ? null : routeDomainPostfix).
-                            endSpec();
-
-                    // removing `expose : true` label from metadata.
-                    removeLabel(routeBuilder.buildMetadata(), EXPOSE_LABEL, "true");
-                    removeLabel(routeBuilder.buildMetadata(), JKubeAnnotations.SERVICE_EXPOSE_URL.value(), "true");
-                    routeBuilder.withNewMetadataLike(routeBuilder.buildMetadata());
-                    routes.add(routeBuilder.build());
-                }
+    private void updateRouteDomainPostFixBasedOnServiceName(String serviceName) {
+        if (StringUtils.isNotBlank(routeDomainPostfix)) {
+            routeDomainPostfix = prepareHostForRoute(routeDomainPostfix, serviceName);
+        } else {
+            routeDomainPostfix = "";
+        }
+    }
+
+    static Route mergeRoute(Route routeFromFragment, Route opinionatedRoute) {
+        // Update ApiVersion to route.openshift.io/v1
+        if (routeFromFragment.getApiVersion().equals("v1")) {
+            routeFromFragment.setApiVersion(opinionatedRoute.getApiVersion());
+        }
+
+        // Merge metadata
```

Comment:
More irrelevant comments, I can read this in the code thanks to the helpful method name `mergeMetadata`
If you do an import static of the method, it will be even more obvious.